sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala

```diff
@@ -36,6 +36,8 @@ import org.apache.spark.util.Utils
 
 
 object DataType {
+  private val curId = new java.util.concurrent.atomic.AtomicLong()
+  def newTypeId = curId.getAndIncrement()
```

Comment:
Now we use the AttributeReference to mark a column, but it is not suitable for nested columns, we need to remove some fields in DataType, it is hard to reconstruct the DataType, so I add a new id to uniquely identify the fields.